src/persist-client/src/internal/trace.rs

```diff
@@ -918,31 +939,190 @@ impl<T: Timestamp + Lattice + Codec64> SpineBatch<T> {
+            original.parts.len()
+        };
+
+        // 2. Replace parts
+        let mut parts = Vec::new();
+        parts.extend_from_slice(&original.parts[..start_part]);
+        parts.extend_from_slice(&replacement.parts);
+        parts.extend_from_slice(&original.parts[end_part..]);
+
+        // 3. Replace run_meta
+        let mut run_meta = Vec::new();
+        run_meta.extend_from_slice(&original.run_meta[..start_run]);
+        run_meta.extend_from_slice(&replacement.run_meta);
+        run_meta.extend_from_slice(&original.run_meta[end_run + 1..]);
+
+        // 4. Rebuild run_splits
+        let mut run_splits = Vec::with_capacity(run_meta.len());
+        let replaced_start = if start_run == 0 {
+            0
+        } else {
+            original.run_splits[start_run - 1]
+        };
+        let replaced_end = if end_run < original.run_splits.len() {
+            original.run_splits[end_run]
+        } else {
+            original.parts.len()
+        };
+        let replaced_len = replaced_end - replaced_start;
+        let replacement_len = replacement.parts.len();
+
+        let prefix = &original.run_splits[..start_run];
+        run_splits.extend_from_slice(prefix);
+
+        let replacement_idx = start_run;
+        let replacement_is_last = replacement_idx + replacement.run_meta.len() == run_meta.len();
+
+        // If we deleted runs, and the replacement is the last run,
+        // we can remove the last split.
+        if replacement.run_meta.is_empty() && replacement_is_last {
+            run_splits.pop();
+        }
+
+        // Only push the replacement split if it's not the final run and the replacement
+        // has runs to add.
+        if !replacement.run_meta.is_empty() && !replacement_is_last {
+            run_splits.push(replaced_start + replacement_len);
+        }
+
+        // 5. Adjust suffix splits
+        if end_run + 1 < original.run_splits.len() {
+            for &split in &original.run_splits[(end_run + 1)..] {
+                let adjusted = split - replaced_len + replacement_len;
+                run_splits.push(adjusted);
+            }
         }
+
+        assert_eq!(
+            run_splits.len(),
+            run_meta.len().saturating_sub(1),
+            "run_splits must have one fewer element than run_meta"
+        );
+
+        let desc = replacement.desc.clone();
+
+        warn!(
+            "recalculating len: original len={}, replaced runs len={} replacement len={}",
+            original.len, replaced_runs_num_updates, replacement.len,
+        );
+        let len = original.len - replaced_runs_num_updates + replacement.len;
+
+        Ok(HollowBatch {
+            desc: desc.clone(),
+            len,
+            parts,
+            run_meta,
+            run_splits,
+        })
     }
 
-    // TODO: Roundtrip the SpineId through FueledMergeReq/FueledMergeRes?
-    /// Checked variant that performs diff sum assertions
     fn maybe_replace_checked<D>(
```

Comment:
This also is creeping up in complexity, but I don't have immediate suggestions... I will play around with it and follow up.